### PR TITLE
[Merged by Bors] - feat: port Order.Hom.Bounded

### PR DESCRIPTION
**Pull request by ChrisHughes24**
mathlib3 SHA : f1a2caaf51ef593799107fe9a8d5e411599f3996

### CODE REVIEW

**Comment by ChrisHughes24** [comment]
There are some Coercion issues here. [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Coercion.20to.20function.20dificulty/near/314675336) discussion

**Comment by Vierkantor** [comment]
If I understand [the conclusions in this thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/FunLike.20issues) correctly, the solution involves replacing all but the final instance parameters in e.g. `BoundedOrderHomClass.toTopHomClass` with implicts, to be solved by unification (this is what the linter is trying to say in its coded way). Let me go ahead and try that.

**Comment by Vierkantor** [comment]
Oh I see why this PR is struggling, in the inheritance chain `OrderIsoClass → BoundedOrderHomClass → TopHomClass` we need a `BoundedOrder` parameter but it only appears in the middle so it cannot be inferred through unification. And we cannot synthesize it because the metavariables haven't been assigned yet.

File: Mathlib/Order/Hom/Bounded.lean [hunk 1]
```diff
@@ -0,0 +1,829 @@
...
+  { ‹BoundedOrderHomClass F α β› with }
+#align bounded_order_hom_class.to_top_hom_class BoundedOrderHomClass.toTopHomClass
+
+-- See note [lower instance priority]
+instance (priority := 100) BoundedOrderHomClass.toBotHomClass {_ : LE α} {_ : LE β}
+    {_ : BoundedOrder α} {_ : BoundedOrder β} [BoundedOrderHomClass F α β] : BotHomClass F α β :=
+  { ‹BoundedOrderHomClass F α β› with }
+#align bounded_order_hom_class.to_bot_hom_class BoundedOrderHomClass.toBotHomClass
+
+@[nolint dangerousInstance] -- The `OrderTop`s should be instance parameters but depend on outParams
```
Comment (Vierkantor) [inline Mathlib/Order/Hom/Bounded.lean:119]: This lint *is* accurate in that this instance can fail to be inferred sometimes, see the `letI`s in `map_eq_top_iff`. Still, there doesn't seem to be another way to make this instance work. See [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Coercion.20to.20function.20dificulty).

**Comment by jcommelin** [comment]
@Vierkantor Do you have a suggestion for the next step?

**Comment by kbuzzard** [comment]
Related Lean 4 issue: https://github.com/leanprover/lean4/issues/1901

**Comment by Vierkantor** [comment]
> Do you have a suggestion for the next step?

Sorry, missed this reply.

Without (significant) changes to Lean 4's inference algorithm I believe this is the best we can do. Note that Lean 3 already struggles in similar conditions, see [mathlib#16919](https://github.com/leanprover-community/mathlib/pull/16919#discussion_r1064521837), and the workaround I propose there will also work here: if we unbundle the inheritance `BoundedOrderHomClass` and `TopHomClass` (or maybe from `BoundedOrderHomClass` to `FunLike`?) we should be able to avoid this issue.

So for now my advice is to merge this PR with the workarounds in place, and hope that we can either revert them when Lean 4 becomes smarter, or not need them due to refactors after the port.

**Pull request closed without merging**